blog-migration.js

```diff
@@ -10,78 +10,109 @@
-import { fetchMarkdown } from './fetch-markdown.js';
-import { getMdast, getTableMap, getNodesByType } from './utils/mdast-utils.js';
-import { saveDocx } from './utils/docx-utils.js';
+import { writeFile, access } from 'fs/promises';
+import { getMdast, getTableMap } from './utils/mdast-utils.js';
+import { saveDocx, saveUpdatedDocx } from './utils/docx-utils.js';
+import { loadMarkdowns, loadIndex } from './utils/fetch-utils.js';
 
 const PROJECT = 'bacom-blog';
 const SITE = 'https://main--business-website--adobe.hlx.page';
 const INDEX = '/blog/query-index.json?limit=3000';
 const USE_CACHE = true;
-const FORCE_SAVE = false;
+const FORCE_SAVE = true;
 
 const MD_DIR = 'md';
 const OUTPUT_DIR = 'output';
+const DOCX_DIR = 'docx';
 const REPORT_DIR = 'reports';
+const MIGRATION = {
+    'pull quote': 'quote',
+    'embed': 'embed',
+    'image': 'image',
+}
 
 export async function main(index, cached, output, force) {
     const reportDir = `${REPORT_DIR}/${PROJECT}`;
     const mdDir = `${MD_DIR}/${PROJECT}`;
+    const docxDir = `${DOCX_DIR}/${PROJECT}`;
     const outputDir = `${output}/${PROJECT}`;
-    const totals = { success: 0, failed: 0 };
-    const failures = [];
-
+    const totals = { succeed: 0, skipped: 0, failed: 0 };
+    const failed = [];
     const indexUrl = `${SITE}${index}`;
-    const mdReport = await fetchMarkdown(PROJECT, SITE, indexUrl, cached);
-
-    console.log('totals', mdReport.totals);
-    console.log('failures', mdReport.failures);
-    const mdReportFile = `${reportDir}/markdown.json`;
-    await mkdir(reportDir, { recursive: true });
-    await writeFile(mdReportFile, JSON.stringify(mdReport, null, 2));
-    console.log(`Report written to ${mdReportFile}`);
-
-    const entries = glob.sync(`./${MD_DIR}/${PROJECT}/**/*.md`);
-    for (let i = 0; i < entries.length; i++) {
-        const entry = entries[i];
+    const entries = await loadIndex(PROJECT, indexUrl, cached);
+    
+    await loadMarkdowns(SITE, mdDir, entries, cached, async (markdown, entry, i) => {
+        if (markdown === null) {
+            console.warn(`Skipping ${entry} as markdown could not be fetched.`);
+            failed.push(entry);
+            totals.failed++;
+            return;
+        }
         try {
-            const markdown = await readFile(entry, 'utf8');
+            const file = entry.split('/').slice(-1)[0];
+            const docxFile = `${file}.docx`;
+            const path = entry.split('/').slice(0, -1).join('/');
+            const sourceFolder = `${docxDir}${path}`;
+            const outputFolder = `${outputDir}${path}`
+            const sourceDocxFile = `${sourceFolder}/${docxFile}`;
+            const outputDocxFile = `${outputFolder}/${docxFile}`;
+
             const mdast = await getMdast(markdown);
+            const tableMap = getTableMap(mdast);
+            if (!tableMap.reduce((acc, block) => {
+                return acc || Object.keys(MIGRATION).includes(block.blockName);
+            }, false)) {
```

Comment:
Nit: This is hard to read in my opinion. Might want to consider breaking it out

Reply:
Thanks, using `tableMap.some` would make this more clear.